Point_set_processing_3/include/CGAL/poisson_eliminate.h

```diff
@@ -0,0 +1,383 @@
+  const bool weight_limiting = parameters::choose_parameter(parameters::get_parameter(np, internal_np::weight_limiting), false);
+  // named parameter for progressive
+  const bool progressive = parameters::choose_parameter(parameters::get_parameter(np, internal_np::progressive), false);
+  // named parameter for tiling
+  const bool tiling = parameters::choose_parameter(parameters::get_parameter(np, internal_np::tiling), false);
+  const unsigned int ambient_dimension = CGAL::Ambient_dimension<Point>::value;
+  const unsigned int dimension = parameters::choose_parameter(parameters::get_parameter(np, internal_np::dimension), 2);
+
+  // named parameter for r_max
+  double r_max = CGAL::to_double(parameters::choose_parameter(parameters::get_parameter(np, internal_np::maximum_radius), 2 * internal::get_maximum_radius(dimension, number_of_points, domain_size)));
+  double r_min = CGAL::to_double(weight_limiting ? internal::get_minimum_radius(points.size(), number_of_points, beta, gamma, r_max) : 0);
+
+  auto weight_functor = parameters::choose_parameter(parameters::get_parameter(np, internal_np::weight_function), internal::Weight_functor<Point>(r_min, alpha));
+
+  std::size_t heap_size = points.size();
+  std::vector<Point> tiling_points;
+
+  IPM ipm(points, tiling_points, point_map);
+
+  auto tile_point = [&tiling_points, &bb, &r_max, &ambient_dimension](const Point& p, std::size_t dim = 0) {
+    auto do_tiling = [&tiling_points, &bb, &r_max, &ambient_dimension](const auto& self, const Point& p, std::size_t dim) -> void {
+      auto it = p.cartesian_begin();
+
+      if (bb.min_coord(int(dim)) > (*(it + dim) - r_max)) {
+        FT v = 2 * bb.min_coord(int(dim)) - (*(it + dim));
+        Point p2;
+        internal::copy_and_replace(p, p2, dim, v);
+        tiling_points.emplace_back(p2);
+        if (dim + 1 < ambient_dimension)
+          self(self, tiling_points.back(), dim + 1);
+      }
+
+      if (bb.max_coord(int(dim)) < (*(it + dim) + r_max)) {
+        FT v = 2 * bb.max_coord(int(dim)) - (*(it + dim));
+        Point p2;
+        internal::copy_and_replace(p, p2, dim, v);
+        tiling_points.emplace_back(p2);
+        if (dim + 1 < ambient_dimension)
+          self(self, tiling_points.back(), dim + 1);
+      }
+
+      if (dim + 1 < ambient_dimension)
+        self(self, p, dim + 1);
+      };
+    do_tiling(do_tiling, p, dim);
+    };
+
+  if (tiling) {
+    for (std::size_t i = 0;i<points.size();i++)
+      tile_point(get(point_map, points[i]));
+  }
+
+  // Tiling requires either to copy the point range or to use a special Index_property_map with a second array of points to keep the input range const.
+  Tree tree(boost::counting_iterator<std::size_t>(0), boost::counting_iterator<std::size_t>(points.size() + tiling_points.size()), Splitter(), Search_traits(ipm));
+  tree.build();
+
+  std::vector<std::size_t> heap(boost::counting_iterator<std::size_t>(0), boost::counting_iterator<std::size_t>(points.size()));
```

Comment:
Do we already have a heap in CGAL? 
Should be use https://www.boost.org/doc/libs/1_86_0/doc/html/heap.html ?
Here I see a `std::vector`  so how do you implemt your heap?

Reply:
I just see `std::make_heap()`  below, so we use something standard.

Reply:
I use std::make_heap for the initial creation. There are also two functions I added: pop_heap and move_down.
In addition, a reverse mapping is kept up-to-date. The heap is implemented in a vector and the size is tracked in the additional variable heap_size, as the indices are kept when progressive ordering is used.

I am not sure that it is easy to switch to an already existing heap implementation or that the code will be shorter.